AC-578 fixing link in label on logistration form

clrux opened the pull request:
# [AC-578](https://openedx.atlassian.net/browse/AC-578)

A recent bug filed indicated that the link for "terms and services" contained within the label on the registration/login form could not be found. This work corrects that by moving the link outside of the label, but resizing things so that the link appears as part of the label.

## Background

Because form fields are navigated by pressing Tab or Up/Down, the link that was contained within the label wasn't accessible. The text was announced, but the link itself wasn't able to be accessed. Because of this, the Terms and Conditions and Privacy information was never able to be viewed by screenreader users prior to registration in the system.

The code now looks something like the following:

```
<input />
<label>I agree to the <span class="sr">Terms and Conditions</span></label>
<div>
<a href>Terms and Conditions</a>
</div>
```

## Screenshot

<img width="678" alt="screen shot 2016-09-12 at 1 34 55 pm" src="https://cloud.githubusercontent.com/assets/2112024/18445899/c8236d44-78ed-11e6-8ddf-003afeeaa379.png">

## Sandbox

[https://clrux-ac-578.sandbox.edx.org](https://clrux-ac-578.sandbox.edx.org/register)

## Future work

Safe lint. Lots of issues. I made [a ticket to correct this](https://openedx.atlassian.net/browse/AC-580) for future work.

## Reviewers

- [x] @cptvitamin 
- [x] @cahrens or @andy-armstrong 
- [x] @lamagnifica FYI

clrux commented:
jenkins run bokchoy

andy-armstrong commented:
@clrux I'm reviewing this as I've recently worked on this API for the mobile app too. I don't think this approach works, as you cannot concatenate localized strings together like this and expect the sentence to still make sense in all languages. See the best practice recommendation against this practice here:

http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/internationalization/i18n.html#general-internationalization-rules

The options I see are to have a single string that is localized with placeholders for the various component pieces, or to break the text into two sentences. Can we leave the API as it was, and just return HTML with the necessary a11y tags pre-inserted?

In addition, it is risky to change this API contract as the mobile client is currently using it and won't work with the changes. I don't like that the API returns HTML because it makes it very specific for the current web client, but that is a trickier thing to change right now.

cahrens commented:
In my opinion, the link should not be part of the control that is rendered via the API. I really dislike the idea of HTML being in the label at all, and as I understand it, it isn't possible to make the link accessible if it is within the label itself.

This is what I would propose-- the API remains as-is (with the caveat that I would like us to move to plain text for the label). The HTML template that is including this control specifies the link in an adjacent element. Something roughly like:

````
<input>
checkbox thing
</input>
<label>
I agree to the terms of service
</label>
<div><a href=link>Click to review the terms of service</link></div>

clrux commented:
@cahrens That's kinda how it is now. Are you suggesting remove the link entirely from the API and place it in the HTML templates?

cahrens commented:
Yes

clrux commented:
Hm. I may set this aside and tackle a few other smaller bugs first.

clrux commented:
@andy-armstrong @cahrens I've made a few adjustments to this. I decided to still keep part of this in the API since it's specific to the register form. But I've simplified it a fair amount. Let me know if this works. Sandbox update in progress.

clrux commented:
@cahrens This one's ready for you. The coverage tests failed, but I'm assuming it's because I ripped out a bunch of tests? How should I go about getting that to pass? Or, how does that work?

cahrens commented:
I'll give it back to @andy-armstrong for review. @andy-armstrong, Chris and I settled on this approach of allowing any form field to have an additional explanatory link. Hopefully that is OK with you. :)

@clrux as far as coveralls, check with testeng.

andy-armstrong commented on lms/static/js/spec/financial-assistance/financial_assistance_form_view_spec.js:
```diff
@@ -25,7 +25,9 @@ define([
                         requiredStr: '',
-                        type: 'select'
+                        type: 'select',
+                        supplementalLink: '',
```
Could you make these new fields optional? It is very annoying to have to add them everywhere.

andy-armstrong commented on lms/templates/student_account/form_field.underscore:
```diff
@@ -4,6 +4,11 @@
             <%= label %>
             <% if ( required && requiredStr ) { %> <%= requiredStr %></label><% } %>
         </label>
+        <% if ( supplementalLink && supplementalText) { %>
```
Nit: lose the space after the opening bracket.

andy-armstrong commented on lms/templates/student_account/form_field.underscore:
```diff
@@ -24,6 +29,11 @@
         <% }); %>
         </select>
         <% if ( instructions ) { %> <span class="tip tip-input" id="<%= form %>-<%= name %>-desc"><%= instructions %></span><% } %>
+        <% if ( supplementalLink && supplementalText) { %>
```
Nit: lose the space after the opening bracket.

andy-armstrong commented on lms/templates/student_account/form_field.underscore:
```diff
@@ -41,6 +51,11 @@
             } %>
             <% if ( required ) { %> aria-required="true" required<% } %> ></textarea>
             <% if ( instructions ) { %> <span class="tip tip-input" id="<%= form %>-<%= name %>-desc"><%= instructions %></span><% } %>
+            <% if ( supplementalLink && supplementalText) { %>
```
Nit: lose the space after the opening bracket.

andy-armstrong commented on lms/templates/student_account/form_field.underscore:
```diff
@@ -65,6 +80,11 @@
             </label>
         <% } %>
         <% if ( instructions ) { %> <span class="tip tip-input" id="<%= form %>-<%= name %>-desc"><%= instructions %></span><% } %>
+        <% if ( supplementalLink && supplementalText) { %>
```
Nit: lose the space after the opening bracket.

andy-armstrong commented on openedx/core/djangoapps/user_api/helpers.py:
```diff
@@ -198,6 +199,12 @@ def add_field(
+                for the form field. An example may be a link to documentation for creating
+                strong passwords.
+
+            supplementalText (unicode): The visible text for the supplemental link avoe.
```
Typo: avoe.

andy-armstrong commented:
@clrux thanks for revisiting this. I like this approach much better, and it is better suited for mobile too in that the API is no longer returning HTML. I have a few minor nits but 👍 from me.

@mdinino this might be relevant to your changes to the Android registration page.

cptvitamin commented:
👍

clrux commented:
@andy-armstrong Thanks for the review. I've addressed your feedback. Let me know if the solution to make the new fields optional is OK. If so, I'll merge if/when tests pass.

clrux commented:
jenkins run bokchoy
